### ISSUE TITLE
Skeleton loader on Cart PMME is absent

### ISSUE THREAD

**Issue by pierorocca** [Automattic/woocommerce-payments#9107]
In #8606 a skeleton was applied to the PMME in order to prevent screen shift and to provide a loading state to the BNPL offer which loads after the page in painted. I'm observing that the skeleton is missing from shortcode and Blocks sites.

**Comment by mdmoore** [comment]
@pierorocca Just tested out the cart on the `develop` branch and I'm seeing the skeleton still in place on cart update. Let me know the site on which it's not working and I'll take a look. 

>I'm observing that the skeleton is missing from shortcode and Blocks sites.

The skeleton is only present on the shortcode cart since the PMME isn't removed from the page in the cart block.

**Comment by pierorocca** [comment]
Thanks @mdmoore. 

Shortcode: https://pieroatomic3.wpcomstaging.com
Blocks: https://proccaatomic.wpcomstaging.com

**Comment by pierorocca** [comment]
https://github.com/user-attachments/assets/9ec400ec-0505-4b4f-aa81-416783f7e7b4


https://github.com/user-attachments/assets/b633fb30-15f0-4c91-824b-105dd48086e6

**Comment by mdmoore** [comment]
Thanks @pierorocca! Was the skeleton loader implemented on PDPs? I think it was only present on shortcode carts. @brettshumaker might know better, but I don't remember ever seeing it on PDPs. I do see the layout shifts on page load though, so maybe this issue can serve as a task to implement skeleton loading on PDPs.

**Comment by pierorocca** [comment]
> Was the skeleton loader implemented on PDPs?

Am I looking for it but can't find it and any PDP related demos are static images. Maybe I missed it but it's a lot more noticeable on the PDP than the Cart no?

**Comment by pierorocca** [comment]
Or maybe when computed styles got applied, it slowed down the paint enough that it's now a noticeable issue?

**Comment by pierorocca** [comment]
> PMME isn't removed from the page in the cart block

True for updates, but on page load? There's a lot of screen shifting happening.

**Comment by FangedParakeet** [comment]
Please [add your planning poker estimate](https://app.zenhub.com/workspaces/Heisenberg-63586f2ec60359001310fa6a/issues/Automattic/woocommerce-payments/9107?planning-poker) with Zenhub @gpressutto5

**Comment by gpressutto5** [comment]
@pierorocca I'm adding the skeleton loader to the blocks cart when it's updated. Does it resolve this issue?

![Jul-25-2024 14-03-09](https://github.com/user-attachments/assets/14ff6c58-7038-4eb8-8fcd-5ae54745d370)

**Comment by pierorocca** [comment]
Needs to be there on page load as well to prevent screen shifting. It's a better UX to not have things moving around.

**Comment by gpressutto5** [comment]
@pierorocca Like in the shortcode cart implementation in #8606, we can't have the skeleton in the initial load because we don't know how large it will be, so we wait for it to be rendered once and save its size. It might not even be displayed. This is out of our control because it is an iframe rendered by Stripe. We could analyze the available data and settings and try to predict whether it is going to be displayed and try to calculate an approximate height, but inevitably, we will have things moving around when we don't get the estimated height correct. It would look like this (the proceed button is moving, but it doesn't look like that because of the flicker):

![Jul-25-2024 15-26-32](https://github.com/user-attachments/assets/93260fcb-9e05-4abe-bd20-0b5839685095)

**Comment by pierorocca** [comment]
Good point. Any other option to get the offer before the page is painted or to make the fetch faster?

**Comment by gpressutto5** [comment]
Unfortunately, no. The PMME is out of our control and can only be loaded in the front end. Samir added an initial loader in #9166. It is similar to the one above, but it doesn't flicker. He also had the same problem of not knowing the correct height and had to estimate it until the component loaded for the first time. Do you think it is better to have this loader with an approximate height or not have the initial loader?

**Comment by pierorocca** [comment]
Correct me if I'm wrong, the PMME will either return nothing because the shopper doesn't qualify (country, currency, total amount) or it will return the logos on the first row and copy on the 2nd row. The copy is wrapped if the container is too narrow. There are also differences between mobile and desktop widths. So there's predictability here of what could be provided in the response.

One approach used by American Eagle is to reserve the max space needed, and if there's an offer it populates otherwise it's left blank.

The other approach I've seen is to shift down once the offer is available with the offer populating a bit more quickly than what I'm seeing from the PMME (are we preconnecting?) and there's only a single, acceptable screen shift down. @FangedParakeet indicated he's trying to minimize shift but that all of the elements we're introducing on the screen are added to the DOM independently so each is causing shift independently. Could you take a look at that work and see if it can be further improved upon?

**Comment by gpressutto5** [comment]
@pierorocca I updated it to use the same logic as Samir used:

![Jul-30-2024 13-54-18](https://github.com/user-attachments/assets/c2952ba4-bb86-4d46-8b6a-52024169e662)

**Comment by pierorocca** [comment]
That's great! @nikkivias for QA. Nikki we're trying to limit screen shifting and buttons and bnpl messaging from jumping around independently of each other. Does the skeleton help or hinder here? It was originally implemented but suffered regression so we've brought it back with tweaks.

**Comment by pierorocca** [comment]
@gpressutto5 could you enabled Google and WooPay please?

**Comment by gpressutto5** [comment]
@pierorocca Sorry, WooPay is not set up in my environment. I recorded this video with Apple Pay! Let me know if you need me to set up WooPay to record a new video.

![Jul-30-2024 18-42-30](https://github.com/user-attachments/assets/be5f2706-e44a-4a89-92a1-e4c86da75c28)

**Comment by pierorocca** [comment]
@FangedParakeet what do you think?

**Comment by pierorocca** [comment]
Here's the current 8.0 experience for comparison:


https://github.com/user-attachments/assets/163824d9-c9ea-468d-bd0b-50da07108083

**Comment by pierorocca** [comment]
@gpressutto5 I think your changes looks better in comparison? Is the slow Apple Pay rendering due to your location or local dev?

**Comment by FangedParakeet** [comment]
> @FangedParakeet what do you think?

Here's what @gpressutto5's changes look like on my local with WooPay & GPay enabled as well. 

https://github.com/user-attachments/assets/278205e5-168b-4df7-9671-5c6aab309a1c

There's a bit of extra spacing around the PMME, but the loading of that specific element looks smooth enough. The express checkout buttons are still making the page leap around quite erratically, but possibly there's potential to implement a similar preloader in a separate issue to improve that.

**Comment by pierorocca** [comment]
Thanks for the video! I slowed it down to .25 speed to get a better idea of what's going on. There's still too much happening even with these improvements.

1. Feedback for the Blocks component/team is the cart component transition is harsh. It doesn't appear to pulse, and then there's too long of a period between when the loader is removed and when the cart is rendered. That makes it feel choppy. The total and proceed to checkout button are also rendered (3rd image) a bit before the cart summary and in the space that will be occupied by the cart summary, adding to the flurry of movement and choppiness. Are we influencing that or is that Blocks behavior?
2. I like that the PMME loader loads at the same time as the cart. Let's make the skeleton occupy the full width of what it could take up. Looks cutoff and unbalanced.
3. Agree on adding a separate issue for the buttons.

![image](https://github.com/user-attachments/assets/472af6bc-8c7d-4683-bc92-ab303bcf28fc)

![image](https://github.com/user-attachments/assets/b92f0232-b069-4a19-be89-3ea4e9ca63fe)

![image](https://github.com/user-attachments/assets/aa0f6201-c99f-413f-a694-628bc8ebc551)

![image](https://github.com/user-attachments/assets/9d2916cf-126b-4edd-b5a6-202bb79ce9cf)

![image](https://github.com/user-attachments/assets/c95da6fe-7e0b-4614-b91c-d1dfa5b90fb1)

![image](https://github.com/user-attachments/assets/1c3d308b-319d-4fa5-92dd-8a4bdc1275b0)

![image](https://github.com/user-attachments/assets/615642d5-9edf-4a33-86e4-771e030c8195)

**Comment by gpressutto5** [comment]
I updated the width and fixed an issue I found with the padding on small screens:

![image](https://github.com/user-attachments/assets/0eef3361-3ca1-4bf5-94b6-672c4a2b475e)

![image](https://github.com/user-attachments/assets/4fa07b32-95b1-4e63-8913-448d6db88e63)

**Comment by pierorocca** [comment]
@nikkivias question to you if we should have a skeleton loader for the PMME and prevent screen shift or ditch the loader and have the PMME smoothly shift the screen down?

**Comment by nikivias** [comment]
@pierorocca  @gpressutto5  Better to use a skeleton loader, engagement is enhanced when users see that content is about to load. However, the skeleton should closely resemble the layout of the fully loaded element to manage expectations correctly.

So something like this, and make sure the height doesn't shift

<img src="https://github.com/user-attachments/assets/23506fe4-eb0c-464d-80c8-929565d24966" width="300px">

**Comment by pierorocca** [comment]
That looks sweet. Ty!

**Comment by gpressutto5** [comment]
@pierorocca @nikkivias We don't know if the body text will have 1 or 2 lines after loading, so I made the bottom loader skeleton slightly bigger. What do you think?


![Aug-07-2024 13-43-37](https://github.com/user-attachments/assets/7b5cde45-a814-46ee-8fe1-c459fccdba18)

![Aug-07-2024 13-44-58](https://github.com/user-attachments/assets/7c76500d-adfc-4f3c-8799-26f42ad12fae)

**Comment by pierorocca** [comment]
True, can you show what it would look like with one line only?

**Comment by gpressutto5** [comment]
@pierorocca The second gif has only one line of text in its body. The first gif has two lines.

**Comment by pierorocca** [comment]
Ah gotcha one line of text plus the line of logos. I believe if there's only 1 logo, both logo and text are on one line?

![image](https://github.com/user-attachments/assets/9accde66-2efd-4c83-a03d-d8ab675ee4aa)

![image](https://github.com/user-attachments/assets/b367bf7b-0dfc-4bd1-82ec-e9d7cb8645d8)

**Comment by gpressutto5** [comment]
![Aug-08-2024 13-30-01](https://github.com/user-attachments/assets/2da477a6-5f31-4c00-b3a6-01f7e30ca2a6)

![Aug-08-2024 13-32-08](https://github.com/user-attachments/assets/28479c47-7148-478a-ba79-b27208d4d23a)

**Comment by pierorocca** [comment]
So net we could have 1 to 3 lines and we don't know in advance, is that an accurate summary? @nikkivias fyi.

**Comment by gpressutto5** [comment]
Yes, that's correct. To be fair, we can potentially have more than 3 lines if the site uses a wide font with a large font size, but that's extremely unlikely as many other components would look broken.

**Comment by nikivias** [comment]
@gpressutto5 Can you tell me what is the minimum and maximum height of the widget once its loaded? I want to create a loader design that will account for both scenarios and anything in between.

**Comment by gpressutto5** [comment]
@nikkivias

TLDR: most likely 12px - 70px, but it can be anything.

We can't say, because it uses the style from the website. Here are some variations I could reproduce using Storefront just by updating the price or the number of methods available, except for the last one. However, there are infinite possibilities by changing the theme or updating the font style (as in the last example).
It can be any size, but most likely >12px as it would be too small to read if it were smaller. As for the maximum height, it might be around 70px, but there's no limit. If the website uses a wide font with a large font size it could be larger than that (as in the last example).

| | Text height | iFrame height (with margin/padding) |
| ---- | ---- | ---- |
| ![image](https://github.com/user-attachments/assets/59a4622c-cfdd-4fc7-aa2f-399c2db1d23c) | 16px | 26px |
| ![image](https://github.com/user-attachments/assets/96bae56b-d07b-42e8-b204-7a2912b12fb0) | 28px | 28px |
| ![image](https://github.com/user-attachments/assets/90fa858b-744a-42e0-aa6d-8e52424749c7) | 36px | 44px |
| ![image](https://github.com/user-attachments/assets/f2986b63-044b-4178-ba2e-9fbb9d7b713a) | 36px | 44px |
| ![image](https://github.com/user-attachments/assets/162e7a2e-cd78-4045-9cc6-e237cad6a0a0) | 39px | 47px |
| ![image](https://github.com/user-attachments/assets/9fd8b307-3bce-499f-9553-758420559f77) | 60px | 68px |
| ![image](https://github.com/user-attachments/assets/dc6008ee-e71c-49a8-88e5-aff3f46cf6bb) | 42px | 50px |
| ![image](https://github.com/user-attachments/assets/25ba0d4f-769a-455c-92dc-bc0acc49eb29) | 60px | 68px |
| ![image](https://github.com/user-attachments/assets/e0c08676-acad-41c0-8986-86187ac2835a) | 68px | 76px |

**Comment by nikivias** [comment]
@gpressutto5 you shared a gif of how it works on Slack:

![Aug-14-2024 14-20-44](https://github.com/user-attachments/assets/f3e71e8c-1420-438f-9aca-02e61523a0a8)

The shifting doesn't look problematic to me because it is alleviated by the smooth motion. Unlike if it were to suddenly just appear, which would be jarring. Since there will be some cases where nothing will be loaded at all (in cases where the cart totals doesn't qualify) I am inclined to propose that we don't use a loader at all. Just need to make sure the appearance of the widget is smooth.

Thoughts? @pierorocca

**Comment by pierorocca** [comment]
> Just need to make sure the appearance of the widget is smooth.

@nikkivias I agree with this. After seeing the loader and not being able to predict the right sizing, it made me rethink if it's needed at all. The smoothness is key so that it feels intentional.

**Comment by gpressutto5** [comment]
@pierorocca @nikkivias

1. We use this same loader on the [shortcode cart](https://github.com/Automattic/woocommerce-payments/issues/8144#issuecomment-2040688316) and [product pages](https://github.com/Automattic/woocommerce-payments/pull/9166). Should we remove them as well? cc @FangedParakeet @mdmoore 
2. In the first implementation, the skeleton loader only appeared when updating the cart after the initial load. Can we leave it without a loader in this case as well?

![Jul-25-2024 14-03-09](https://github.com/user-attachments/assets/14ff6c58-7038-4eb8-8fcd-5ae54745d370)

**Comment by pierorocca** [comment]
If it's smooth we can follow the same approach.

On refresh, similarly if we can keep it super smooth and prevent screen shifting, then I'm ok removing the loader.

@nikkivias ?

**Comment by gpressutto5** [comment]
Without a skeleton loader it just updates the text.

![Aug-15-2024 12-30-48](https://github.com/user-attachments/assets/23dfc28c-81b4-4e7b-aa0b-5836aa71cc9d)

**Comment by pierorocca** [comment]
Looks pretty seamless to me and updates in the same way as tax, totals, etc. do. so it's all consistent.
